src/storage/src/hummock/sstable/block_iterator.rs

```diff
@@ -246,7 +261,10 @@ impl BlockIterator {
     }
 
     /// Searches the restart point index that the given `key` belongs to.
-    fn search_restart_point_index_by_key(&self, key: FullKey<&[u8]>) -> usize {
+    fn search_restart_point_index_by_key(&mut self, key: FullKey<&[u8]>) -> usize {
+        // Create a temporary local hitmap.
```

Comment:
Why not just update the local hitmap instead of creating a new one?

Reply:
The block binary search requires the iterator's immutable reference, and the hitmap cannot be updated in the closure for it requires the mutable reference. (I forgot to restore this diff line, let me fix it.)

Reply:
Looks like the immutable reference is required to call `decode_prefix_at` of `self`. However, looking into the implementation of `decode_prefix_at`, it only uses `self.block`. Therefore, if we change to implement `decode_prefix_at` as method of `BlockHolder`, we don't need the immutable reference of `self` here. And it makes sense to me to make `decode_prefix_at` a method of a single block instead of a method of `BlockIterator`.

Reply:
Or we can just inline the call of `decode_prefix_at` as
```
KeyPrefix::decode(
    &mut &self.block.data()[offset..],
    offset,
    key_len_type,
    value_len_type,
)
```

Reply:
Looks like the `offset` information is duplicated here, once in slice range, and once as method parameter. 

Maybe we can change the first parameter of `decode` from `buf: &mut impl Buf` to `&[u8]` and we generate the `&mut &self.block.data()[offset..]` inside the code of `KeyPrefix::decode` so that we remove the duplicated offset information in method call.

Reply:
+1 for inline call, but I think duplicated `offset` sounds okay here, it will unify the encode/decode parameters of `KeyPrefix`.